### ISSUE TITLE
Why Python and Java RS behave differently for some simulations?

### ISSUE THREAD

**Issue by akolonin** [singnet/reputation#135]
@deborahduong - please provide details to reproduce the issue with runnable code in github and run instructions for single instance of market conditions and reputation system parameters

**Comment by akolonin** [comment]
Here is the code that I run:
https://github.com/singnet/reputation/blob/master/agency/python/src/performance_test.py

Here is the output
$ python performance_test.py
{'default': 0.5, 'decayed': 0.5, 'conservatism': 0.5, 'precision': 0.01, 'weighting': True, 'fullnorm': True, 'liquid': False, 'logranks': False, 'logratings': False, 'downrating': False, 'update_period': 1, 'aggregation': False}
Good: 136336213 Bad: 34109852 Good to Bad 23938198 Good/Bad: 3.996974627740982 Bad/Good_to_Bad: 1.4249131033171336 LTS: 0.17558209571216415 PFS: 0.7017971816471089
Time spent Aigents: 1664.3371043205261
{'default': 0.5, 'conservatism': 0.5, 'precision': 0.01, 'weighting': True, 'fullnorm': True, 'liquid': False, 'logranks': False, 'aggregation': False, 'logratings': False, 'update_period': 1, 'use_ratings': True, 'date': datetime.date(2018, 1, 1)}
Good: 136441481 Bad: 34111210 Good to Bad 26443749 Good/Bad: 3.999901527972769 Bad/Good_to_Bad: 1.2899536294948193 LTS: 0.19381018738722133 PFS: 0.7752216646668353
Time spent Python: 168.68842315673828
What I see is that 'decayed' and 'aggregation' properties are not set. I have run only 31 days so that is why LTS & PFS may be close enough. @nejc9921 - please make sure that all the settings are set as expected. If needed, please extend the unit tests for decayed and aggregation to cover the case, or let me know if I am missing something or need my involvement.

**Comment by akolonin** [comment]
@nejc9921 in your PR I still don't see 'downrating' returned https://github.com/singnet/reputation/pull/137/files - please fix.

**Comment by nejc9921** [comment]
Some bug fixes. Fixed return in get_parameters. There was also a missed bug in getting ranks from days other than previous day - then we get wrong values.

However, after running performance_test, there is still an existing problem, so this did not solve it. Perhaps this will help in comparison to development of Java code;  the development of rating for agent '500' in Python:
reputations saved by date 2018-01-01 {'id': '500', 'rank': 86.0}
reputations saved by date 2018-01-02 {'id': '500', 'rank': 75.0}
reputations saved by date 2018-01-03 {'id': '500', 'rank': 69.0}
reputations saved by date 2018-01-04 {'id': '500', 'rank': 43.0}
reputations saved by date 2018-01-05 {'id': '500', 'rank': 62.0}
reputations saved by date 2018-01-06 {'id': '500', 'rank': 42.0}
reputations saved by date 2018-01-07 {'id': '500', 'rank': 47.0}
reputations saved by date 2018-01-08 {'id': '500', 'rank': 41.0}
reputations saved by date 2018-01-09 {'id': '500', 'rank': 34.0}
reputations saved by date 2018-01-10 {'id': '500', 'rank': 46.0}
reputations saved by date 2018-01-11 {'id': '500', 'rank': 45.0}
reputations saved by date 2018-01-12 {'id': '500', 'rank': 40.0}
reputations saved by date 2018-01-13 {'id': '500', 'rank': 47.0}
reputations saved by date 2018-01-14 {'id': '500', 'rank': 66.0}
reputations saved by date 2018-01-15 {'id': '500', 'rank': 63.0}
reputations saved by date 2018-01-16 {'id': '500', 'rank': 41.0}
reputations saved by date 2018-01-17 {'id': '500', 'rank': 48.0}
reputations saved by date 2018-01-18 {'id': '500', 'rank': 55.0}
reputations saved by date 2018-01-19 {'id': '500', 'rank': 64.0}
reputations saved by date 2018-01-20 {'id': '500', 'rank': 56.0}
reputations saved by date 2018-01-21 {'id': '500', 'rank': 54.0}
reputations saved by date 2018-01-22 {'id': '500', 'rank': 61.0}
reputations saved by date 2018-01-23 {'id': '500', 'rank': 38.0}
reputations saved by date 2018-01-24 {'id': '500', 'rank': 47.0}
reputations saved by date 2018-01-25 {'id': '500', 'rank': 38.0}
reputations saved by date 2018-01-26 {'id': '500', 'rank': 54.0}
reputations saved by date 2018-01-27 {'id': '500', 'rank': 53.0}
reputations saved by date 2018-01-28 {'id': '500', 'rank': 68.0}

**Comment by akolonin** [comment]
@nejc9921 - this may be caused by #157

**Comment by akolonin** [comment]
Script and results in "/home/akolonin/aigents/reputation/reputation":
test_params_10.py
test_params_10_java.txt
test_params_10_python.txt

Discrepancy:
{'default': 0.0, 'decayed': 0.5, 'conservatism': 0.5, 'precision': 0.01, 'weighting': True, 'denomination': True, 'fullnorm': True, 'liquid': True, 'logranks': True, 'logratings': True, 'downrating': False, 'update_period': 1, 'aggregation': False, 'unrated': True, 'ratings': 1.0, 'spendings': 0.0} Good: 22281 Bad: 4470 Good to Bad: 3637 Good/Bad: 4.984563758389262 Bad/Good_to_Bad: 1.2290349188891945 LTS: 0.16323324805888426 PFS: 0.8136465324384787
{'default': 0.0, 'decayed': 0.5, 'conservatism': 0.9, 'precision': 0.01, 'weighting': True, 'denomination': True, 'fullnorm': True, 'liquid': True, 'logranks': True, 'logratings': False, 'downrating': False, 'update_period': 1, 'aggregation': False, 'unrated': True, 'ratings': 1.0, 'spendings': 0.0} Good: 21775 Bad: 5259 Good to Bad: 1185 Good/Bad: 4.140521011599163 Bad/Good_to_Bad: 4.4379746835443035 LTS: 0.05442020665901263 PFS: 0.2253280091272105
vs
{'default': 0.0, 'conservatism': 0.5, 'precision': 0.01, 'weighting': True, 'fullnorm': True, 'liquid': True, 'logranks': True, 'aggregation': False, 'logratings': True, 'update_period': 1, 'use_ratings': True, 'date': datetime.date(2018, 1, 1), 'decayed': 0.5, 'downrating': False, 'denomination': True, 'unrated': True} Good: 22961 Bad: 6124 Good to Bad: 1185 Good/Bad: 3.749346832135859 Bad/Good_to_Bad: 5.167932489451477 LTS: 0.05160925046818518 PFS: 0.1935009797517962
{'default': 0.0, 'conservatism': 0.9, 'precision': 0.01, 'weighting': True, 'fullnorm': True, 'liquid': True, 'logranks': True, 'aggregation': False, 'logratings': False, 'update_period': 1, 'use_ratings': True, 'date': datetime.date(2018, 1, 1), 'decayed': 0.5, 'downrating': False, 'denomination': True, 'unrated': True} Good: 21786 Bad: 5155 Good to Bad: 2121 Good/Bad: 4.226188166828322 Bad/Good_to_Bad: 2.4304573314474305 LTS: 0.0973561002478656 PFS: 0.4114451988360815

**Comment by nejc9921** [comment]
new issue found. This one is small, however in  test_params_10.py it produced results which diverged a lot. The thing is about rounding the ratings. How it works:
Python:
Say agent 1 has rating of 0.866583373. get_ratings() method will return 87 (rounded). When calculating ranking updates for the next day, it will use the number 0.866583373 and build on that number.

Java:
Say agent 1 has rating of 0.866583373. get_ratings() method will return 87 (rounded). When calculating ranking updates for the next day, it will use the number 0.87 and build on that number.

The differences in unittest are very small and that is why they often avoided me. What is worse is that I am not sure if they stay like that - the ratings in both systems later on even converge after being slightly wrong in 1 period if time.

One would think that this should have very limited impact on ratings, however as soon as one rating is rounded differently, whole set of transactions change and all the subsequent ratings change. An example is given [here](https://docs.google.com/spreadsheets/d/1b2gH3T_Aizl402pyA13KHRL79U7erpEg-vI0d5N7yDQ/edit#gid=802416481).  There you can see simulated agents of one system on one side and the other on the other side. On the red, there is a set of transactions which changes rounded numbers on merely 2 agents. In the yellow you see how the systems start diverging in transactions right after that.

More can be found [here](https://docs.google.com/document/d/1jN06angSsvLe424Sk5uO8ib9nsOse0OOwwaC_xP3vGA/edit#heading=h.gjdgxs). On first page you can see how ranks differ if we have the same dataset. You can see that there was a slight mistake on date: 2018-01-03, however it was fixed later on somehow.

If you scroll down this document, you can see that if we generate different agents, however, this does not automatically fix because differentials are different and therefore ranks are completely different then. It shows that reputation system is highly sensitive in such a case, but also that rounding sometimes differs between Java and Python.

**Comment by nejc9921** [comment]
ok, made a new push which fixed the roundings issue about which I talked in a post above. The issue was found by running test_params_10.py and inspecting the differences between Java and Python. There were two issues found, I focused on one and hoped that solution on one would solve the other one as well. It did solve the first issue, but the second issue still exists. So, there are still some unexplained deviations.

[Here ](https://docs.google.com/document/d/1frMo8i5XssEiRYCn7MSQBdEvlOgKqy9tZduhBGoDaw4/edit)is the document comparing Python and Java runs. We can still find one deviation that I have to inspect.

**Comment by nejc9921** [comment]
Further discoveries on issues that make the Java and Python systems deviate:
Issue 1: First of all, we have strange parameters:'default':0.0,'decayed':0.5, 'conservatism':0.9}.
We get the following values for differential: {'2': 0.0, '3': 0.0, '4': 0.0, '9': 0.0}.
As one can see, those values are then transferred to ratings for both Java and Python. The difference happens with rating givers, so buyers. I can say what happens in Python (have not yet inspected Java code): They start with default of 0 and then they go towards 0.5 with conservativity of 0.9. So, 0.5*0.9 + 0.1 * 0. In Java it’s just the opposite. I just changed and all the tests are still passing. A new unittest was created (test_decay_buyers) and with new fixes, both Python and Java are now passing the test (old code will not pass the test).

Issue 2: Maybe more interesting one (because I failed to fix it):
I added a closer inspection of the problem in test_downratings2. I believe this is the problem with downratings. What happens is that given the set of parameters, we have the following ranks:
Date 1:
Python ranks: {'2': 0.0, '3': 0.0, '4': 0.0, '9': 0.0, '5': 25.0, '6': 25.0, '7': 25.0, '8': 25.0, '10': 25.0}
Java ranks: {'10': 25.0, '5': 25.0, '6': 25.0, '7': 25.0, '8': 25.0, '2': 0.0, '3': 0.0, '4': 0.0, '9': 0.0}

Date 2:
Python ranks: {'2': 50.0, '3': 52.0, '4': 100.0, '9': 0.0, '5': 75.0, '6': 75.0, '7': 75.0, '8': 75.0, '10': 75.0}
Java ranks: {'10': 100.0, '5': 100.0, '6': 100.0, '7': 100.0, '8': 100.0, '2': 68.0, '3': 0.0, '4': 0.0, '9': 0.0}

In the second day, the ratings deviate strongly. Unittest captures that. I can go into detail about what happens in Python in the update;
First, we get differential:
{'4': 2.5312205808879145, '9': -2.7722550101886716, '3': 0.0}
Negative value makes sense, because one of the ratings to agent 9 had rating of 0 and since downratings=True, we transform it to -1. We then normalize this (fullnorm=True):
{'4': 1.0, '9': 0.0, '3': 0.5227241952151446}
So, we get the feeling how the updates will look. Since conservatism = 0.5, we update it to:
{'2': 0.25, '3': 0.2613620976075723, '4': 0.5, '9': 0.0, '5': 0.375, '6': 0.375, '7': 0.375, '8': 0.375, '10': 0.375}
And normalize it to:
{'2': 0.5, '3': 0.5227241952151446, '4': 1.0, '9': 0.0, '5': 0.75, '6': 0.75, '7': 0.75, '8': 0.75, '10': 0.75}
Thus getting the Python ranks on date 2 as I copy-pasted above.
If we look at the ranks for agents 3,4 and 9 in Java, we can see that they all get set up with a value of 0. I am not sure why this is the case and I will try to figure it out. I can take a closer look at Java code tomorrow for better inspection.

**Comment by akolonin** [comment]
After few more fixes, there is still minor discrepancies running: https://github.com/singnet/reputation/blob/master/reputation/reputation_scenario_test.py
All the differences are pretty small, so it could be just matter of an inconsistent rounding or incorrect determination of the ranges.

Also, don't understand while same things are apparently done in redundant way, like I have commented in https://github.com/singnet/reputation/commit/f1a7572dcc1b81685842c61bd771411f24867a80#diff-ed0153d75772c87f70f1ad3c1d398885R259

@nejc9921 - I suggest you to review the code carefully, eliminate all pieces and execution branches that are working now and follow the same order of transformations like it is found in the reference implementation: https://github.com/aigents/aigents-java/blob/master/src/main/java/net/webstructor/peer/Reputationer.java#L344

**Comment by akolonin** [comment]
# Here is the code that reproduce that:

import time

import datetime
from reputation_scenario import reputation_simulate
from reputation_service_api import *
from aigents_reputation_api import AigentsAPIReputationService

def dict_sorted(d):
	first = True
	s = "{"
	for key, value in sorted(d.items(), key=lambda x: x[0]): 
		template = "'{}': {}" if first else ", '{}': {}"
		s += template.format(key, value)
		first = False
	s += "}"
	return s

#TODO use any other Reputation Service here
rs = None

verbose = False
days = 10 #10
consumers = 0.5
suppliers = 0.5
good_range = [1,8]
bad_range = [9,10]

good_transactions = 1
bad_transactions = 1

#Trying different RS parameters
good_agent = {"range": good_range, "values": [100,1000], "transactions": good_transactions, "suppliers": suppliers, "consumers": consumers}
bad_agent = {"range": bad_range, "values": [100,1000], "transactions": bad_transactions, "suppliers": suppliers, "consumers": consumers}
print('Good Agent:',str(good_agent))
print('Bad Agent :',str(bad_agent))
sp = 4
sip = 0 # sp/2

def test():
	"""
	This one is different:
anton-mac:reputation demo$ python discrepancy_something_test_10.py
Good Agent: {'range': [1, 8], 'values': [100, 1000], 'transactions': 1, 'suppliers': 0.5, 'consumers': 0.5}
Bad Agent : {'range': [9, 10], 'values': [100, 1000], 'transactions': 1, 'suppliers': 0.5, 'consumers': 0.5}
Java:
{'aggregation': False, 'conservatism': 0.7, 'decayed': 0.5, 'default': 0.0, 'denomination': True, 'downrating': False, 'fullnorm': True, 'liquid': True, 'logranks': True, 'logratings': False, 'precision': 0.01, 'ratings': 0.5, 'spendings': 0.5, 'unrated': False, 'update_period': 1, 'weighting': True} Good: 22323 Bad: 4043 Good to Bad: 640 Good/Bad: 5.521395003710116 Bad/Good_to_Bad: 6.3171875 LTS: 0.028669981633293016 PFS: 0.15829829334652487
Python:
{'aggregation': False, 'conservatism': 0.7, 'decayed': 0.5, 'default': 0.0, 'denomination': True, 'downrating': False, 'fullnorm': True, 'liquid': True, 'logranks': True, 'logratings': False, 'precision': 0.01, 'ratings': 0.5, 'spendings': 0.5, 'unrated': False, 'update_period': 1, 'weighting': True} Good: 22547 Bad: 5281 Good to Bad: 640 Good/Bad: 4.26945654232153 Bad/Good_to_Bad: 8.2515625 LTS: 0.028385151017873776 PFS: 0.12118916871804583
	"""
	rs.set_parameters({'fullnorm':True,'weighting':True ,'logratings':False,'downrating':False,'denomination':True ,'unrated':False,'default':0.0,'decayed':0.5,'ratings':0.5,'spendings':0.5,'conservatism':0.7})
	print(dict_sorted(rs.get_parameters()), end=" ")
	reputation_simulate(good_agent,bad_agent, datetime.date(2018, 1, 1), days, True, rs, campaign = [sp,sip], verbose=False)
	

rs = AigentsAPIReputationService('http://localtest.com:1180/', 'john@doe.org', 'q', 'a', False, 'test', True)
print('Java:')
test()
rs = PythonReputationService()
print('Python:')
test()

**Comment by nejc9921** [comment]
I found the issue and made a unittest with the differences. The meat of the issue is quite simple. There are 5 transactions and after update_ranks() method, they get different ranks. Java and Python seem to agree that ratings play no role here - the only thing that moves ranks are spendings.

Spendings are calculated differently. I suspect the difference is in normalization, but it could as well be in calculation of spendings. I will go through exact process here;
Relevant code:
self.assertEqual(rs.put_ratings([{'from':5,'type':'rating','to':2,'value':0.25,'weight':682,'time':dt2}]),0)
		self.assertEqual(rs.put_ratings([{'from':6,'type':'rating','to':3,'value':1,'weight':220,'time':dt2}]),0)
		self.assertEqual(rs.put_ratings([{'from':7,'type':'rating','to':4,'value':1,'weight':583,'time':dt2}]),0)
		self.assertEqual(rs.put_ratings([{'from':8,'type':'rating','to':2,'value':1,'weight':196,'time':dt2}]),0)
		self.assertEqual(rs.put_ratings([{'from':10,'type':'rating','to':9,'value':1,'weight':129,'time':dt2}]),0)
		rs.set_parameters({'fullnorm':True,'weighting':True ,'logratings':False,'downrating':False,'denomination':True ,'unrated':False,'default':0.0,'decayed':0.5,'ratings':0.9,'spendings':0.1,'conservatism':0.5})

Python gets the following ranks:
{'2': 0.0, '5': 100.0, '3': 0.0, '6': 16.0, '4': 0.0, '7': 82.0, '8': 12.0, '9': 0.0, '10': 0.0}
And Java:
{'5': 100.0, '7': 91.0, '6': 32.0, '8': 25.0, '10': 0.0, '2': 0.0, '3': 0.0, '4': 0.0, '9': 0.0}

I am again not sure how Java gets those numbers. But for Python, it goes this way:
First, spendings get summed; 
 {'5': 68200.0, '6': 22000.0, '7': 58300.0, '8': 19600.0, '10': 12900.0}
Then, they get normalized (fullnorm=True):
{'5': 1.0, '6': 0.16455696202531644, '7': 0.8209764918625678, '8': 0.12115732368896925, '10': 0.0}

and then we more or less get the numbers that I mentioned. I failed to find out how Java behaves, so currently I am stuck at this point.


Also, there are still many deviations. This problem describes one and will solve at least 1 of 7 outstanding problems still there. Hopefully it will solve the rest as well.

**Comment by akolonin** [comment]
There are two minor discrepancies remaining for running 1000 agents 183 days in "space exploration" mode https://github.com/singnet/reputation/blob/master/reputation/reputation_scenario_test.py - comparing Java vs Python output below:
akolonin@Ubuntu-1604-xenial-64-minimal:~/aigents/reputation/reputation$ diff discrepancy_search_1000_j.txt discrepancy_search_1000_p.txt 
5c5
< Regular RS: Good: 86009099 Bad: 4530398 Good to Bad: 1727749 Good/Bad: 18.984888082680595 Bad/Good_to_Bad: 2.622138979678182 LTS: 0.020087979296237016 PFS: 0.3813680387462647
---
> Regular RS: Good: 86002435 Bad: 4528508 Good to Bad: 1721377 Good/Bad: 18.991339973342214 Bad/Good_to_Bad: 2.63074736097903 LTS: 0.020015444911530702 PFS: 0.380120119032582
11c11
< Regular RS: Good: 86080727 Bad: 4489520 Good to Bad: 2023767 Good/Bad: 19.17370387034694 Bad/Good_to_Bad: 2.218397671273422 LTS: 0.023510105810328485 PFS: 0.45077580676776136
---
> Regular RS: Good: 86051206 Bad: 4492149 Good to Bad: 2035243 Good/Bad: 19.155910901441604 Bad/Good_to_Bad: 2.207180665895915 LTS: 0.023651533715866806 PFS: 0.4530666725435866

**Comment by akolonin** [comment]
I just have run "discrepancy test" on 10, 1K and 10K agents (with my latest PR). The test for 1K has shown 2 (of 16) discrepancies under 1%, while the tests for 10 and 10K are all identical across 16 different conditions/configurations - see the test in my latest PR: 
https://github.com/singnet/reputation/pull/226

@nejc9921 - are you sure you latest PR: 
https://github.com/singnet/reputation/pull/228
makes it even better, can you run it for 10K test before we can merge it?
